jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,505 @@
+        # mixdepth from which payment is sourced
+        assert isinstance(mixdepth, int)
+        self.mixdepth = mixdepth
+        assert isinstance(destination, btc.CCoinAddress)
+        self.destination = destination
+        assert isinstance(amount, int)
+        assert amount > 0
+        self.amount = amount
+        self.server = server
+        self.pj_state = self.JM_PJ_INIT
+        self.payment_tx = None
+        self.initial_psbt = None
+        self.payjoin_psbt = None
+        self.final_psbt = None
+        # change is initialized as None
+        # in case there is no change:
+        self.change_out = None
+
+    def set_payment_tx_and_psbt(self, in_psbt):
+        assert isinstance(in_psbt, btc.PartiallySignedTransaction)
+        self.initial_psbt = in_psbt
+        # any failure here is a coding error, as it is fully
+        # under our control.
+        assert self.sanity_check_initial_payment()
+        self.pj_state = self.JM_PJ_PAYMENT_CREATED
+
+    def sanity_check_initial_payment(self):
+        """ These checks are motivated by the checks specified
+        for the *receiver* in the btcpayserver implementation doc.
+        We want to make sure our payment isn't rejected.
+        We also sanity check that the payment details match
+        the initialization of this Manager object.
+        """
+        # failure to extract tx should throw an error;
+        # this PSBT must be finalized and sane.
+        self.payment_tx = self.initial_psbt.extract_transaction()
+
+        # inputs must all have witness utxo populated
+        for inp in self.initial_psbt.inputs:
+            if not inp.utxo and isinstance(inp.utxo, btc.CTxOut):
+                return False
+
+        # check that there is no xpub or derivation info
+        if self.initial_psbt.xpubs:
+            return False
+        for inp in self.initial_psbt.inputs:
+            # derivation_map is an OrderedDict, if empty
+            # it will be counted as false:
+            if inp.derivation_map:
+                return False
+        for out in self.initial_psbt.outputs:
+            if out.derivation_map:
+                return False
+
+        # TODO we can replicate the mempool check here for
+        # Core versions sufficiently high, also encapsulate
+        # it in bc_interface.
+
+        found_amt = 0
+        found_addr = 0
+        # our logic requires no more than one change output
+        # for now:
+        assert len(self.payment_tx.vout) in [1, 2]
+        for out in self.payment_tx.vout:
+            if out.nValue == self.amount:
+                found_amt += 1
+            if btc.CCoinAddress.from_scriptPubKey(
+                out.scriptPubKey) == self.destination:
+                found_addr += 1
+            else:
+                # store this for our balance check
+                # for receiver proposal
+                self.change_out = out
+        if not found_amt == 1 and found_addr == 1:
```

Comment:
I think you want to say `not (found_amt == 1 and found_addr == 1)` here, `and` has lower precedence than `not`

Reply:
Also while I don't really understand the context here because this is foreign code to me, but it seems to me that there could be situation where `found_amt == 1 and found_addr == 1` will be true when the change output matches against the amount on first iteration, while the other output matches against the scriptPubKey on the second iteration. This may be intentional of course, as I said I don't understand the context.

Reply:
Oh what a terrible mistake, thanks!
I guess I was having a bad day (hour?) yesterday when I wrote that, because the logic isn't clean anyway, even correcting that syntax error: we should identify our payment output by address *and* amount together, not separately (else the payment amount could be in the change output!) Doh!

Reply:
Oh it seems you already asked about the same thing while I was replying :) Fix incoming.